rice-datanode/source/data_transfer_server.cc

```diff
@@ -212,20 +213,22 @@ void TransferServer::processReadRequest(tcp::socket& sock) {
 	}
 
 	uint64_t blockID = proto.header().baseheader().block().blockid();
-	std::string block;
-	bool success = fs->getBlock(blockID, block);
+	nativefs::block_info info;
+	bool success = fs->getBlockInfo(blockID, info);
 	if (!success) {
-		LOG(ERROR) << "Failure on fs.getBlock";
+		LOG(ERROR) << "Failure on fs.getBlockInfo";
 	}
 
 	uint64_t offset = proto.offset();
-	uint64_t len = std::min(block.size() - offset, proto.len());
-	if (offset > block.size()) {
+	uint64_t len = std::min(info.len - offset, proto.len());
+	if (offset > info.len) {
 		len = 0;
 	}
 	bool send_chksums = proto.sendchecksums();
 
 	uint64_t seq = 0;
+	std::string payload;
+	std::lock_guard<std::mutex> lock(readFS);
```

Comment:
I'm not totally decided on whether the lock is required, your call. I'm not sure which is heavier: potentially waiting on slow clients to finish reading or waiting on disk to seek around.